File: 0205-isomorphic-strings/0205-isomorphic-strings.cpp
class Solution {
public:
    bool isIsomorphic(string s, string t) {
        int n = s.length();
        int m = t.length();

        unordered_map<char, char> umap1;
        
        unordered_map<char, char> umap2;

        if(n != m){
            return false;
        }

        for( int i = 0; i< n; i++){


            if(umap1.count(s[i]) && umap1[s[i]] != t[i]){
                return false;
            }
            if(umap2.count(t[i]) && umap2[t[i]] != s[i]){
                return false;
            }
            
            umap1[s[i]] = t[i];
            umap2[t[i]] = s[i];

        }

        return true;
    }
};

File: 0205-isomorphic-strings/0205-isomorphic-strings.py
class Solution:
    # 7. Isomorphic Strings (LeetCode 205)
    def isIsomorphic(self, s: str, t: str) -> bool:
        # Use two dictionaries to map characters from s to t and t to s respectively
        map_s_t, map_t_s = {}, {}
        
        for c1, c2 in zip(s, t):
            if (c1 in map_s_t and map_s_t[c1] != c2) or (c2 in map_t_s and map_t_s[c2] != c1):
                return False
            map_s_t[c1] = c2
            map_t_s[c2] = c1
            
        return True